### DISCUSSION TITLE
How to handle DRC exclusions in KiCad 6

### DISCUSSION THREAD

**Discussion by cdwilson** [INTI-CMNB/KiBot#230] [General]
I've got a design with a large castellated mounting hole:

<img width="1231" alt="image" src="https://user-images.githubusercontent.com/14631/179857677-b2ce01e6-94bb-4140-b0ad-a10ecd97c78e.png">

This castellated mounting hole is created by "slicing" the board edge line (on`Edge.Cuts`)  through a normal mounting hole footprint.  However, this creates multiple DRC violations that I need to "Exclude" in KiCad 6 DRC GUI:

<img width="1304" alt="image" src="https://user-images.githubusercontent.com/14631/179858228-036f4604-16bc-490e-9d13-cfb09880387f.png">

These exclusions are stored in the KiCad project file (`*.kicad_pro`):

```json
      "drc_exclusions" : [
        "copper_edge_clearance|142250000|98000000|00000000-0000-0000-0000-000060704a77|4b8a595c-9b48-4157-8b12-0a90675da7aa",
        "copper_edge_clearance|142250000|98000000|00000000-0000-0000-0000-000060704a77|59f8ee19-dd93-46ab-9aa6-2685af0b6909",
        "copper_edge_clearance|142250000|98000000|00000000-0000-0000-0000-000060704a77|c8916864-fb8d-4f6f-be56-7b73db75bb64",
        "copper_edge_clearance|142482331|98876227|00000000-0000-0000-0000-000060704a95|0b5e6c81-8708-44e7-8204-5cd8847f3e76",
        "copper_edge_clearance|144000000|99752454|00000000-0000-0000-0000-000060704a95|d6d19ec8-bc97-47d1-a730-a9486e913b4a",
        "copper_edge_clearance|145517669|98876227|00000000-0000-0000-0000-000060704a95|12b8f71b-2075-4810-8da4-4cfdd67db115"
      ],
```

From what I can tell, KiBot doesn't use these exclusions, so I have to filter these errors in the preflight section of the KiBot config:

```
preflight:
  filters:
    - filter: 'Ignore board edge clearance violation for castellated hole on J1'
      error: 'copper_edge_clearance'
      regex: '(?s)Edge.Cuts(.*)J1'
```

However, when I run KiBot on this design, it clears out the `drc_exclusions` list in the KiCad project file:

<img width="1452" alt="image" src="https://user-images.githubusercontent.com/14631/179858949-ae6fcc0d-c56f-4aa3-add0-ff997586c817.png">

I keep the design in version control, so I have to run `git checkout my_project.kicad_pro` each time after running KiBot to restore the exclusions.  It's a bit inconvenient, so I'm wondering if there is a way to make KiBot work better with KiCad 6 DRC exclusions (not sure if there is a better way to handle this).

Question 1: why does KiBot delete the DRC exclusions in the KiCad project file?

Question 2: is there a way to tell KiBot to use the existing DRC exclusions already set in the KiCad project file? (rather than having to define separate filters in the KiBot config?)

**Comment by set-soft** [comment]
Hi @cdwilson !

> Question 1: why does KiBot delete the DRC exclusions in the KiCad project file?

Is not KiBot, is KiCad itself. The Python API doesn't currently handle the exclusions. Is reported as a bug. And it looks like the API is saving the project at some point that KiBot doesn't expect it.
 
> Question 2: is there a way to tell KiBot to use the existing DRC exclusions already set in the KiCad project file? (rather than having to define separate filters in the KiBot config?)

Currently we don't have a mechanism, is in my todo list, but I want to finish other stuff first. And I'm not sure we can handle it, plus it shouldn't be necesary if KiCad is fixed.

**Reply by cdwilson** [reply]
@set-soft thanks for the explanation.  

> The Python API doesn't currently handle the exclusions. Is reported as a bug.

Is this the bug you were referring to?  https://gitlab.com/kicad/code/kicad/-/issues/11562

**Reply by set-soft** [reply]
Yes
